README.md
# Creates QR codes images 

This library is fork of Sam Vincent rqrcode-rails3 https://github.com/samvincent/rqrcode-rails3
Dependes on mini_magic and [rqrcode](https://github.com/whomwah/rqrcode) 

This gem supports generating QR images in SVG, BMP, TIF, JPG, GIF and PNG format. 

## Installation

Add the following to your +Gemfile+.

    gem 'qrcoder'

or 

    gem install qrcoder

## How to use

### Image method 

If you use image method you must define output format. 

Allowed formats are 
  
  * png 
  * bmp 
  * svg 
  * jpg
  * gif 
  * tif

    QRCode.image(text, output, options)

Examples 

    QRCode.image("some text", "/home/user/", :format => [:png, :svg], :filename => "simple_test" 
    QRCode.image("some text", "/home/user/", :format => :png, :filename => "simple_test" , :unit => 12

### SVG output
  
Return svg output 

    QRCode.svg(text, output, options)

Example

    QRCode.svg("some text", "/home/user", :filename => "simple_test", :unit => 12) 

### Other methods

You can use other methods
  
    QRCode.bmp(text, output, options)
    QRCode.png(text, output, options)
    QRCode.jpg(text, output, options)
    QRCode.tif(text, output, options)
    QRCode.gif(text, output, options)
  
#### Options:

* `:size`   – This controls how big the QR Code will be. The smallest size will be chosen by default. Set to maintain consistent size.
* `:level`  – The error correction level, can be:
  * Level `:l` 7%  of code can be restored
  * Level `:m` 15% of code can be restored
  * Level `:q` 25% of code can be restored
  * Level `:h` 30% of code can be restored (default :h) 
* `:offset` – Padding around the QR Code (e.g. 10)
* `:unit`   – How many pixels per module (e.g. 11)
* `:fill`   – Background color (e.g "ffffff" or :white)
* `:color`  – Foreground color for the code (e.g. "000000" or :black)
  
## About

This library is fork of Sam Vincent [rqrcode-rails3](https://github.com/samvincent/rqrcode-rails3)

QR codes are encoded by [rqrcode](https://github.com/whomwah/rqrcode)


History
== 0.1.0 2012-05-07
* Initial version


test/rqrcode_svg_test.rb
require "test/unit"
require_relative "../lib/qrcoder.rb"

class QRCoder::QRCodeTest <Test::Unit::TestCase
  include QRCoder
  CURRENT_DIR = File.dirname(File.expand_path(__FILE__)) 

  def test_creating_png 
    output_file = CURRENT_DIR + "/new_test.png" 
    QRCode.image("test", CURRENT_DIR, { :format => :png , :filename => "new_test" })
    assert(File.exist?(output_file))
    File.delete(output_file)
  end
  
  def test_creating_png_with_larger  
    output_file = [{:size => "small", :unit => 11}, {:size => "medium" ,:unit => 20}, {:size => "large",:unit => 41}]
    output_file.each do |file|
      QRCode.image("test", CURRENT_DIR,  :format => :png , :filename => file[:size], :unit => file[:unit] )
      path = CURRENT_DIR + "/" + file[:size] + ".png"
      assert(File.exist?(path))
      image = MiniMagick::Image.open(path)
      assert_equal(image[:width], file[:unit] * 21) 
      File.delete(path)
    end
  end

  def test_creating_jpg 
    output_file = CURRENT_DIR + "/new_test.jpg" 
    QRCode.image("test", CURRENT_DIR, { :format => :jpg , :filename => "new_test", :unit => 40 })
    assert(File.exist?(output_file))
    File.delete(output_file)
  end
  
  def test_creating_svg 
    output_file = CURRENT_DIR + "/new_test.svg" 
    QRCode.image("tesfadsghjklghjklfghjklfghrgthjkmlfghjktghjkfghjkffghjkl", CURRENT_DIR, { :format => :svg , :filename => "new_test" })
    assert(File.exist?(output_file))
    File.delete(output_file)
  end

  def test_creating_multiple 
    formats = [:svg, :png, :jpg, :gif, :tif, :bmp]
    QRCode.image("tesfadsghjjklfghrgthjkmlfghjktghjkfghjkffghjkl", CURRENT_DIR, { :format => formats , :filename => "mulitple_test" })
    formats.each do |format|
      output_file = CURRENT_DIR + "/mulitple_test.#{format}" 
      assert(File.exist?(output_file))
      File.delete(output_file)
    end
  end

  def test_invalid_file_format 
    assert_raise QRCoder::QRCodeError do  
      QRCode.image("ghjk", CURRENT_DIR, {:format => :rtyui})
    end
  end
  
  def test_invalid_file_path 
    assert_raise QRCoder::QRCodeError do  
      QRCode.image("ghjk", "rtyufhadisuyiuds hfioh isufh", {:format => :png})
    end
  end

  def test_other_methods 
    formats = [:png, :jpg, :gif, :tif, :bmp]
    QRCode.bmp("tesfajkfghjkffghjkl", CURRENT_DIR, {:filename => "meta_test" })
    QRCode.png("tesfajkfghjkffghjkl", CURRENT_DIR, {:filename => "meta_test" })
    QRCode.jpg("tesfajkfghjkffghjkl", CURRENT_DIR, {:filename => "meta_test" })
    QRCode.gif("tesfajkfghjkffghjkl", CURRENT_DIR, {:filename => "meta_test" })
    QRCode.tif("tesfajkfghjkffghjkl", CURRENT_DIR, {:filename => "meta_test" })

    formats.each do |format|
      output_file = CURRENT_DIR + "/meta_test.#{format}" 
      assert(File.exist?(output_file))
      File.delete(output_file)
    end
  end

  
end


lib/rqrcoder/renderers/svg.rb
module QRCoder
  module Renderers
    class SVG
      class << self
        # Render the SVG from the qrcode string provided from the RQRCode gem
        #   Options:
        #   offset - Padding around the QR Code (e.g. 10)
        #   unit   - How many pixels per module (Default: 11)
        #   fill   - Background color (e.g "ffffff" or :white)
        #   color  - Foreground color for the code (e.g. "000000" or :black)

        def render(qrcode, options={})
          offset  = options[:offset].to_i || 0
          color   = options[:color]       || "000"
          unit    = options[:unit]        || 11

          # height and width dependent on offset and QR complexity
          dimension = (qrcode.module_count*unit) + (2*offset)

          xml_tag   = %{<?xml version="1.0" standalone="yes"?>}
          open_tag  = %{<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:ev="http://www.w3.org/2001/xml-events" width="#{dimension}" height="#{dimension}">}
          close_tag = "</svg>"

          result = []
          qrcode.modules.each_index do |c|
            tmp = []
            qrcode.modules.each_index do |r|
              y = c*unit + offset
              x = r*unit + offset

              next unless qrcode.is_dark(c, r)
              tmp << %{<rect width="#{unit}" height="#{unit}" x="#{x}" y="#{y}" style="fill:##{color}"/>}
            end 
            result << tmp.join
          end
          
          if options[:fill]
            result.unshift %{<rect width="#{dimension}" height="#{dimension}" x="0" y="0" style="fill:##{options[:fill]}"/>}
          end
          
          svg = [xml_tag, open_tag, result, close_tag].flatten.join("\n")
        end
      end
    end
  end
end


lib/qrcoder/size_calculator.rb
module QRCoder
  module SizeCalculator
    # size - seems to follow this logic
    #     # | input | modules 
    #       | size  | created
    #-------|-------|--------
    #     1 |     7 |      21
    #     2 |    14 |      25 (+4)
    #     3 |    24 |      29   -
    #     4 |    34 |      33   -
    #     5 |    44 |      37   -
    #     6 |    58 |      41   -
    #     7 |    64 |      45   -
    #     8 |    84 |      49   -
    #     9 |    98 |      53   -
    #    10 |   119 |      57   -
    #    11 |   137 |      61   -
    #    12 |   155 |      65   -
    #    13 |   177 |      69   -
    #    14 |   194 |      73   -
    
    QR_CHAR_SIZE_VS_SIZE = [7, 14, 24, 34, 44, 58, 64, 84, 98, 119, 137, 155, 177, 194]
    
    def minimum_qr_size_from_string(string)
      QR_CHAR_SIZE_VS_SIZE.each_with_index do |size, index|
        return (index + 1) if string.size < size
      end
      
      # If it's particularly big, we'll try and create codes until it accepts
      i = QR_CHAR_SIZE_VS_SIZE.size
      begin
        i += 1
        QRCoder::QRCode.new(string, :size => i)
        return i
      rescue QRCoder::QRCodeRunTimeError
        retry
      end
    end
  end
end


lib/qrcoder.rb
require 'rqrcode'
require 'qrcoder/size_calculator.rb'
require 'qrcoder/renderers/svg.rb'
require 'mini_magick'

module QRCoder
  extend SizeCalculator

  class QRCodeError < StandardError; end

  class QRCode 
    IMAGE_FORMATS = [:jpg , :png, :gif, :tif, :bmp]

    class << self

      # 
      # Create qr image or images 
      #
      # == Parameters
      #
      # * text    <tt>String</tt> -- String for qrcode
      # * output  <tt>String</tt> -- String for path where to save files
      # * options <tt>Hash</tt>   -- Hash with options
      #
      # === Options  
      #
      # * :format   - Array or symbol. Allowed formats are :png, :bmp, :svg, :jpg, :gif and :tif
      # * :filename - String for filename without extension. Extension will be added automaticly.
      # * :size     - This controls how big the QR Code will be. Smallest size will be chosen by default. Set to maintain consistent size.
      # * :level    - The error correction level, can be:
      #   * Level :l 7%  of code can be restored
      #   * Level :m 15% of code can be restored
      #   * Level :q 25% of code can be restored
      #   * Level :h 30% of code can be restored (default :h) 
      # * :offset - Padding around the QR Code (e.g. 10)
      # * :unit   - How many pixels per module (e.g. 11)
      # * :fill   - Background color (e.g "ffffff" or :white)
      # * :color  - Foreground color for the code (e.g. "000000" or :black)
      #
      # === Example 
      # 
      #  QRCode.image("some text", "/home/user/", :format => [:png, :svg], :filename => "simple_test" 
      #  QRCode.image("some text", "/home/user/", :format => :png, :filename => "simple_test" , :unit => 12
      #
      def image text, output, options
        formats = []
        if options[:format].kind_of? Array
          formats = options[:format]
        else
          formats << options[:format] || :png
        end
        filename = options[:filename] || text
        svg      = self.svg(text, options)

        err_f = formats - IMAGE_FORMATS - [:svg]
        raise QRCodeError.new("#{err_f.join(', ')} is not supported file formats") unless(err_f.empty?)

        begin 
          formats.each do |format|
            if IMAGE_FORMATS.include?(format)
              image = MiniMagick::Image.read(svg) { |i| i.format "svg" }
              image.format format.to_s
              image.write "#{output}/#{filename}.#{format}"
            else
              data = svg
              file = File.new("#{output}/#{filename}.#{format}", "w")
              file.write(data)
              file.close
            end
          end
        rescue 
          raise QRCodeError.new "File saving error"
        end
      end

      # Create qr code svg output response 
      # * text    <tt>String</tt> -- String for qrcode
      # * options <tt>Hash</tt>   -- Hash with options
      #
      # === Options  
      #
      # * :format   - Array or symbol. Allowed formats are :png, :bmp, :svg, :jpg, :gif and :tif
      # * :filename - String for filename without extension. Extension will be added automaticly.
      # * :size     - This controls how big the QR Code will be. Smallest size will be chosen by default. Set to maintain consistent size.
      # * :level    - The error correction level, can be:
      #   * Level :l 7%  of code can be restored
      #   * Level :m 15% of code can be restored
      #   * Level :q 25% of code can be restored
      #   * Level :h 30% of code can be restored (default :h) 
      # * :offset - Padding around the QR Code (e.g. 10)
      # * :unit   - How many pixels per module (e.g. 11)
      # * :fill   - Background color (e.g "ffffff" or :white)
      # * :color  - Foreground color for the code (e.g. "000000" or :black)
      #
      def svg text, options ={}
        size   = options[:size]  || QRCoder.minimum_qr_size_from_string(text)
        level  = options[:level] || :h
        qrcode = RQRCode::QRCode.new(text, :size => size, :level => level)
        QRCoder::Renderers::SVG::render(qrcode, options)
      end


      IMAGE_FORMATS.each do |format| 
        define_method format.to_s do |text, output, options|
          options[:format] = [format]
          self.image text, output, options
        end
      end
    end
  end

end
